ui/org.eclipse.pde.core/src/org/eclipse/pde/internal/core/ClasspathComputer.java

```diff
@@ -237,45 +255,63 @@ protected static IBuild getBuild(IProject project) throws CoreException {
 		return (buildModel != null) ? buildModel.getBuild() : null;
 	}
 
-	private static void addLibraryEntry(IProject project, IPluginLibrary library, IPath sourceAttachment, IClasspathAttribute[] attrs, ArrayList<IClasspathEntry> result) throws JavaModelException {
+	private static void addLibraryEntry(IPluginLibrary library, Map<String, IPath> sourceLibraryMap,
+			ClasspathConfiguration context) throws JavaModelException {
 		String name = ClasspathUtilCore.expandLibraryName(library.getName());
-		IResource jarFile = project.findMember(name);
+		IResource jarFile = context.javaProject.getProject().findMember(name);
 		if (jarFile == null) {
 			return;
 		}
 
-		IPackageFragmentRoot root = JavaCore.create(project).getPackageFragmentRoot(jarFile);
+		IPath sourceAttachment = sourceLibraryMap.get(library.getName());
+		boolean isExported = library.isExported();
+
+		IPackageFragmentRoot root = context.javaProject.getPackageFragmentRoot(jarFile);
 		if (root.exists() && root.getKind() == IPackageFragmentRoot.K_BINARY) {
 			IClasspathEntry oldEntry = root.getRawClasspathEntry();
-			// If we have the same binary root but new or different source, we should recreate the entry
-			if ((sourceAttachment == null && oldEntry.getSourceAttachmentPath() != null) || (sourceAttachment != null && sourceAttachment.equals(oldEntry.getSourceAttachmentPath()))) {
-				if (!result.contains(oldEntry)) {
-					result.add(oldEntry);
-					return;
-				}
+			// If we have the same binary root but new or different source,
+			// we should recreate the entry, preserving unrelated attributes.
+			if (sourceAttachment == null) {
+				// no override: stick to existing one, if any
+				sourceAttachment = oldEntry.getSourceAttachmentPath();
 			}
+			if (sourceAttachment != null && sourceAttachment.equals(oldEntry.getSourceAttachmentPath())) {
```

Comment:
Can't we keep the original code here?
```suggestion
			if ((sourceAttachment == null && oldEntry.getSourceAttachmentPath() != null) || (sourceAttachment != null && sourceAttachment.equals(oldEntry.getSourceAttachmentPath()))) {
```

Reply:
If I remember correctly, I have had issues around this original code because it was ignorant against the (new==null && old==null) situation, but this really was because the subsequent code wasn't preserving some attributes.
So indeed, combined with the other fixes, the original code here makes sense again.